demos/coreMQTT_Agent/simple_sub_pub_demo.c

```diff
@@ -0,0 +1,560 @@
+    xApplicationDefinedContext.xTaskToNotify = xTaskGetCurrentTaskHandle();
+    xApplicationDefinedContext.pArgs = ( void * ) &xSubscribeArgs;
+
+    xCommandParams.blockTimeMs = mqttexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = prvSubscribeCommandCallback;
+    xCommandParams.pCmdCompleteCallbackContext = ( void * ) &xApplicationDefinedContext;
+
+    /* Loop in case the queue used to communicate with the MQTT agent is full and
+     * attempts to post to it time out.  The queue will not become full if the
+     * priority of the MQTT agent task is higher than the priority of the task
+     * calling this function. */
+    LogInfo( ( "Sending subscribe request to agent for topic filter: %s with id %d",
+               pcTopicFilter,
+               ( int ) ulSubscribeMessageID ) );
+
+    do
+    {
+        /* TODO: prvIncomingPublish as publish callback. */
+        xCommandAdded = MQTTAgent_Subscribe( &xGlobalMqttAgentContext,
+                                             &xSubscribeArgs,
+                                             &xCommandParams );
+    } while( xCommandAdded != MQTTSuccess );
+
+    /* Wait for acks to the subscribe message - this is optional but done here
+     * so the code below can check the notification sent by the callback matches
+     * the ulNextSubscribeMessageID value set in the context above. */
+    xCommandAcknowledged = prvWaitForCommandAcknowledgment( NULL );
+
+    /* Check both ways the status was passed back just for demonstration
+     * purposes. */
+    if( ( xCommandAcknowledged != pdTRUE ) ||
+        ( xApplicationDefinedContext.xReturnStatus != MQTTSuccess ) )
+    {
+        LogInfo( ( "Error or timed out waiting for ack to subscribe message topic %s",
+                   pcTopicFilter ) );
+    }
+    else
+    {
+        LogInfo( ( "Received subscribe ack for topic %s containing ID %d",
+                   pcTopicFilter,
+                   ( int ) xApplicationDefinedContext.ulNotificationValue ) );
+    }
+
+    return xCommandAcknowledged;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvSimpleSubscribePublishTask( void * pvParameters )
+{
+    MQTTPublishInfo_t xPublishInfo = { 0 };
+    char payloadBuf[ mqttexampleSTRING_BUFFER_LENGTH ];
+    char taskName[ mqttexampleSTRING_BUFFER_LENGTH ];
+    MQTTAgentCommandContext_t xCommandContext;
+    uint32_t ulNotification = 0U, ulValueToNotify = 0UL;
+    MQTTStatus_t xCommandAdded;
+    struct DemoParams * pxParams = ( struct DemoParams * ) pvParameters;
+    uint32_t ulTaskNumber = pxParams->ulTaskNumber;
+    MQTTQoS_t xQoS;
+    TickType_t xTicksToDelay;
+    MQTTAgentCommandInfo_t xCommandParams = { 0 };
+    char * pcTopicBuffer = topicBuf[ ulTaskNumber ];
+    uint32_t numSuccesses = 0U;
+
+    /* Have different tasks use different QoS.  0 and 1.  2 can also be used
+     * if supported by the broker. */
+    xQoS = ( MQTTQoS_t ) ( ulTaskNumber % 2UL );
```

Comment:
Should the `2UL` be a configurable macro to allow the demo to be tested with Qos2 if the broker supports it?